# Final Project

### 1. Description

The goal of this project assignment is to develop a data analytics project using a real-world
dataset. Students should select a public dataset, think about a research project, for example,
t-test, two-sample hypothesis tests, regression models, classification, clustering, or any data
analytic task, implement the data model (using libraries or without libraries), evaluate, and
interpret the results. You can apply more than one data model.

### 2. Public Dataset

Select a data set from the available public datasets.
You can find a list of such public data sets here:
http://www.teymourian.de/public-data-sets-for-data-analytic-projects/)
- The data set must not be very Large.
- You can reuse one of the assignment data sets.

In [159]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, classification_report

In [160]:
df = pd.read_csv('lap_times.csv')

# df['driverRef'].value_counts

df.head()

,raceId,driverId,lap,position,time,milliseconds
0,841,20,1,1,1:38.109,98109
1,841,20,2,1,1:33.006,93006
2,841,20,3,1,1:32.713,92713
3,841,20,4,1,1:32.803,92803
4,841,20,5,1,1:32.342,92342


In [161]:
df = pd.read_csv('qualifying.csv')
df

,qualifyId,raceId,driverId,constructorId,number,position,q1,q2,q3
0,1,18,1,1,22,1,1:26.572,1:25.187,1:26.714
1,2,18,9,2,4,2,1:26.103,1:25.315,1:26.869
2,3,18,5,1,23,3,1:25.664,1:25.452,1:27.079
3,4,18,13,6,2,4,1:25.994,1:25.691,1:27.178
4,5,18,2,2,3,5,1:25.960,1:25.518,1:27.236
...,...,...,...,...,...,...,...,...,...
10489,10547,1144,848,3,23,16,1:23.821,\N,\N
10490,10548,1144,855,15,24,17,1:23.880,\N,\N
10491,10549,1144,1,131,44,18,1:23.887,\N,\N
10492,10550,1144,861,3,43,19,1:23.912,\N,\N


In [162]:
df = pd.read_csv('races.csv')
years = np.sort(df['year'].unique())
print(years)
names = np.sort(df['name'].unique())
print(names)

[1950 1951 1952 1953 1954 1955 1956 1957 1958 1959 1960 1961 1962 1963
 1964 1965 1966 1967 1968 1969 1970 1971 1972 1973 1974 1975 1976 1977
 1978 1979 1980 1981 1982 1983 1984 1985 1986 1987 1988 1989 1990 1991
 1992 1993 1994 1995 1996 1997 1998 1999 2000 2001 2002 2003 2004 2005
 2006 2007 2008 2009 2010 2011 2012 2013 2014 2015 2016 2017 2018 2019
 2020 2021 2022 2023 2024]
['70th Anniversary Grand Prix' 'Abu Dhabi Grand Prix'
 'Argentine Grand Prix' 'Australian Grand Prix' 'Austrian Grand Prix'
 'Azerbaijan Grand Prix' 'Bahrain Grand Prix' 'Belgian Grand Prix'
 'Brazilian Grand Prix' 'British Grand Prix' 'Caesars Palace Grand Prix'
 'Canadian Grand Prix' 'Chinese Grand Prix' 'Dallas Grand Prix'
 'Detroit Grand Prix' 'Dutch Grand Prix' 'Eifel Grand Prix'
 'Emilia Romagna Grand Prix' 'European Grand Prix' 'French Grand Prix'
 'German Grand Prix' 'Hungarian Grand Prix' 'Indian Grand Prix'
 'Indianapolis 500' 'Italian Grand Prix' 'Japanese Grand Prix'
 'Korean Grand Prix' 'Las Vegas 

In [163]:
df = pd.read_csv('results.csv')
df

,resultId,raceId,driverId,constructorId,number,grid,position,positionText,positionOrder,points,laps,time,milliseconds,fastestLap,rank,fastestLapTime,fastestLapSpeed,statusId
0,1,18,1,1,22,1,1,1,1,10.0,58,1:34:50.616,5690616,39,2,1:27.452,218.300,1
1,2,18,2,2,3,5,2,2,2,8.0,58,+5.478,5696094,41,3,1:27.739,217.586,1
2,3,18,3,3,7,7,3,3,3,6.0,58,+8.163,5698779,41,5,1:28.090,216.719,1
3,4,18,4,4,5,11,4,4,4,5.0,58,+17.181,5707797,58,7,1:28.603,215.464,1
4,5,18,5,1,23,3,5,5,5,4.0,58,+18.014,5708630,43,1,1:27.418,218.385,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
26754,26760,1144,825,210,20,14,16,16,16,0.0,57,\N,\N,57,1,1:25.637,222.002,11
26755,26761,1144,859,215,30,12,17,17,17,0.0,55,\N,\N,52,12,1:28.751,214.212,5
26756,26762,1144,822,15,77,9,\N,R,18,0.0,30,\N,\N,14,19,1:29.482,212.462,130
26757,26763,1144,861,3,43,20,\N,R,19,0.0,26,\N,\N,5,17,1:29.411,212.631,5


In [164]:
df = pd.read_csv('sprint_results.csv')
df

,resultId,raceId,driverId,constructorId,number,grid,position,positionText,positionOrder,points,laps,time,milliseconds,fastestLap,fastestLapTime,statusId
0,1,1061,830,9,33,2,1,1,1,3,17,25:38.426,1538426,14,1:30.013,1
1,2,1061,1,131,44,1,2,2,2,2,17,+1.430,1539856,17,1:29.937,1
2,3,1061,822,131,77,3,3,3,3,1,17,+7.502,1545928,17,1:29.958,1
3,4,1061,844,6,16,4,4,4,4,0,17,+11.278,1549704,16,1:30.163,1
4,5,1061,846,1,4,6,5,5,5,0,17,+24.111,1562537,16,1:30.566,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
355,356,1143,859,215,30,10,16,16,16,0,19,+34.356,1657366,18,1:25.762,1
356,357,1143,852,215,22,16,17,17,17,0,19,+35.102,1658112,17,1:25.838,1
357,358,1143,861,3,43,0,18,18,18,0,19,+35.639,1658649,18,1:25.599,1
358,359,1143,855,15,24,18,19,19,19,0,19,+1:11.436,1694446,19,1:25.051,1


### 3. Define a Project

Define a data science research question based on the selected dataset. You can select one of
the listed datasets or search on the web, and pick up one of the available public datasets.
You should describe the following items
1. What is your data set about?
2. Clean up your data and reduce it to no more than 2000 observations if your data set is
very large.
3. What is exactly your research question? What do you want to learn from data? What
is your learning model ,e.g., a Classification, Clustering, etc?
4. What is your current expectations about the results?
5. How do you want to evaluate your project? How to access the correctness of your
model? How well would you expect that the model will work?

### work
1. This dataset is about ... f1 race data and results
2. clean up dataset. data from races 1950-2024 (up to 1144 entries in each dataset depending on how far back data was collected)
3. Research Question:

Use various ML models to predict if drivers score points (top 10 finishers) or no points (bottom 10 finishers) in a race. At any time, there are up to 20 drivers on the grid, 10 teams competing, 2 drivers per team (no more, no less).

Goal of ML model is Classification. Class '0' means no points scored, class '1' means points scored.

May also try out some clustering methods and validate results to see if the model is any good (eg. Kmeans and DBScan clustering)

4. Current expectations of results: ...

5. How to evaluate project... 

metrics (accuracy, precision, recall, f1-score), AUC/ROC curve

### 4. Implementation

- You need to implement your project in python. You are allowed to use any Machine
Learning Library like scikit-learn1. You are also allowed to implement your project
without using any libraries.
- You can use Jupyter Notebooks to implement your project and provide your documen-
tations.
- You code should be completed and be compilable without any errors. We should be
able to read your documents, and be able to run your project.
- Run your implementation (on your Laptop or on Cluster if data is large) and generate
the results.
- Provide the interpretations of your results.
- What can you do to improve your results? Apply your ideas to improve your results.
- Provide any references that you have used.

In [ ]:
# ideas
# ROC curve and AUC, compare: LogReg, KNN, RandomForest, (and maybe SVM/kernels, DBScan)
# predict if a driver will finish with points for a given race (places <=10 class '1', else class '0')
# validation/metrics: accuracy, f-score, precision/recall, ROC/AUC

### 5. Create Recorded Presentation Video of your Project

- Create a document or a presentation (You can create a presentation using your Jupyter
notebook, or create a powerpoint presentation, or other formats) to describe your
project and results
- Describe your code.
- Describe the results of your project in a professional way.
- You may want to visualization diagrams and describe the results based on some diagrams but having diagrams is not a MUST have to get the full credit.
- Describe the model and results of your project in a way that every person in this field
can read, enjoy and understand.
Record your presentation in form of a video recording. Present your talk on your laptop
computer and record it using a desktop recording software.
You should use your webcam so that we can identify who is presenter.

**<u>Note 1</u>**: Your video presentation should be between 8 minutes to maximum 15 minutes
long (Min 8 mi, Max 15 min)
**<u>Note 3</u>**: You should turn on your Web Camera so that we can see your face during the
presentation, and be able to identify the presenter.
Grading will be based on quality of your presentation, and correct describing of algorithms
or concepts.

#### 5.1 Turnins

- Submit your Jupyter Notebook or Python code
- Submit your documentation (It can be part of your Jupyter Notebook or a separate file
like PPT or Docs, or other formats)
- Submit your Video Presentation. Upload your video file to Canvas Panopto Video
Gallery.

#### 5.2 

Your term project will be graded based on the following rubrics:
1. Correct description of the selected dataset.
2. Correctness of your data preparation and cleanup approach.
3. General Correctness of data analysis approach.
4. Correctness of your implementation and results.
5. Correctness of your interpretations of the results.
6. Providing all references.

In [ ]:
# ideas

# cluster drivers by performance (unsupervised learning)
# goal - group drivers based on race performance results
# features - average lap time, pit stops, qualifying position, points, race results
# kmeans clustering, hierarchical clustering
# evaluation - silhouette score, cluster interpretability (visualization)

# driver points prediction
# target: binary (1 if position <= 10, else 0)
# features: qualifying position, pit stops, lap times, previous standings, constructor performance, circuit type, weather (if available)
# models: LogReg, Decision tree, random forest, gradient boosting
# evaluation: accuracy, precision/recall, f1-score, ROC/AUC